crates/pathfinder/src/sequencer/reply.rs

```diff
@@ -237,41 +237,74 @@ pub mod transaction {
     }
 
     /// Represents deserialized L2 transaction data.
-    ///
-    /// TODO refactor into a 3-variant enum (Declare, Deploy, Invoke)
+    #[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+    #[serde(untagged)]
+    #[serde(deny_unknown_fields)]
+    pub enum Transaction {
+        Declare(DeclareTransaction),
+        Deploy(DeployTransaction),
+        Invoke(InvokeTransaction),
+    }
+
+    impl Transaction {
+        /// Returns hash of the transaction
+        pub fn hash(&self) -> StarknetTransactionHash {
+            match self {
+                Transaction::Declare(t) => t.transaction_hash,
+                Transaction::Deploy(t) => t.transaction_hash,
+                Transaction::Invoke(t) => t.transaction_hash,
+            }
+        }
+    }
+
+    /// Represents deserialized L2 declare transaction data.
     #[serde_as]
     #[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
     #[serde(deny_unknown_fields)]
-    pub struct Transaction {
-        #[serde_as(as = "Option<Vec<CallParamAsDecimalStr>>")]
+    pub struct DeclareTransaction {
+        pub class_hash: ClassHash,
+        #[serde_as(as = "FeeAsHexStr")]
+        pub max_fee: Fee,
+        pub nonce: TransactionNonce,
+        pub sender_address: ContractAddress,
+        #[serde_as(as = "Vec<TransactionSignatureElemAsDecimalStr>")]
         #[serde(default)]
-        pub calldata: Option<Vec<CallParam>>,
-        /// None for Invoke, Some() for Deploy and Declare
+        pub signature: Vec<TransactionSignatureElem>,
+        pub transaction_hash: StarknetTransactionHash,
+        pub r#type: Type,
+        #[serde_as(as = "TransactionVersionAsHexStr")]
+        pub version: TransactionVersion,
+    }
+
+    /// Represents deserialized L2 deploy transaction data.
+    #[serde_as]
+    #[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+    #[serde(deny_unknown_fields)]
+    pub struct DeployTransaction {
+        pub contract_address: ContractAddress,
+        pub contract_address_salt: ContractAddressSalt,
         #[serde(default)]
         pub class_hash: Option<ClassHash>,
```

Comment:
I made this field optional because of what I noticed in cairo cli code, but now I think this just does not make sense.

Reply:
I don't think it makes sense either. I've now checked the transaction JSON dumps in my Goerli database and there's not a single DEPLOY transaction where `class_hash` is not present.